File: day_06/gbm_py/day_06.py
from math import prod, sqrt, floor, ceil

# def num_winnable(time, dist):
#     winnable = 0
#     for held_time in range(0, time + 1):
#         if (time - held_time) * held_time > dist:
#             winnable += 1
#     return winnable
#


def num_winnable(time, dist):
    d = (time**2) - (4 * -1 * -dist)
    x1 = (-time + sqrt(d)) / -2
    x2 = (-time - sqrt(d)) / -2
    return ceil(x2) - floor(x1) - 1


with open("input.txt", "rt") as infile:
    times = [int(t) for t in infile.readline().split(':')[1].split()]
    dists = [int(d) for d in infile.readline().split(':')[1].split()]

winnables = list()
for t, d in zip(times, dists):
    winnables.append(num_winnable(t, d))

print(f'Part 1: {prod(winnables)}')

time = int("".join([str(t) for t in times]))
dist = int("".join([str(d) for d in dists]))

print(f'Part 2: {num_winnable(time,dist)}')

File: day_06/gbm_lua/day_06.lua
local infile = assert(io.open("input.txt", "r"))
io.input(infile)

local function line_as_list()
    local line = io.read("*line")
    local items = {}
    for item in line:gmatch("%d*") do table.insert(items, tonumber(item)) end
    return items
end

local times = line_as_list()
local dists = line_as_list()

local function num_winning(time, dist)
    local d = (time * time) - (4 * -1 * -dist)
    local x1 = (-time + math.sqrt(d)) / -2
    local x2 = (-time - math.sqrt(d)) / -2
    return math.ceil(x2) - math.floor(x1) - 1
end

local part_1_prod = 1
for i, _ in pairs(times) do
    part_1_prod = part_1_prod * num_winning(times[i], dists[i])
end

local part_2 = num_winning(tonumber(table.concat(times)), tonumber(table.concat(dists)))

print(string.format("Part 1: %d", part_1_prod))
print(string.format("Part 2: %d", part_2))

